sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -609,8 +610,10 @@ abstract class TypeCoercionBase {
       case c @ Concat(children) if conf.concatBinaryAsString ||
         !children.map(_.dataType).forall(_ == BinaryType) =>
         val newChildren = c.children.map { e =>
-          implicitCast(e, StringType).getOrElse(e)
+          if (e.dataType.isInstanceOf[StringType]) e
+          else implicitCast(e, StringType).getOrElse(e)
```

Comment:
We should update `def implicitCast` if we don't want to allow implicit cast `_: StringType` to `StringType`